### ISSUE TITLE
Add Meta.unapply(Queries)

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9488]
jOOQ 3.13 has introduced `Meta.apply(Queries)` which runs a set of queries on an interpreter, based on a previously known `Meta` snapshot: https://github.com/jOOQ/jOOQ/issues/9436.

Most queries that we're currently interpreting have an obvious inverse query, for example:

```sql
CREATE SCHEMA s;
DROP SCHEMA s;

ALTER TABLE t ADD (i int, j int);
ALTER TABLE t DROP (i, j);

ALTER TABLE t RENAME COLUMN c1 TO c2;
ALTER TABLE t RENAME COLUMN c2 TO c1;
```

It should be possible to implement `Meta.unapply()` (or a different name) by traversing the argument queries in the inverse way, and reversing them, to achieve, for any `Meta` and any `Queries`:

```java
Meta m = ...;
Queries q = ...;

assertEquals(m, m.apply(q).unapply(q));
```

The two most popular database change management products in the Java ecosystem (Flyway and Liquibase) support undoing migrations. Liquibase can do it "easily" for XML based migrations, but neither can do it for plain SQL migrations. We could provide a tool / plugin for both of those products, to parse and reverse the forward migration, in order to produce the undo migration for SQL scripts.

This only works for a given set of DDL statements. We cannot undo arbitrary `UPDATE` statements, for example, nor restore lost data if tables or columns are dropped.

**Comment by knutwannheden** [comment]
That sounds cool. I assume you would use `Meta::diff` to implement this, which then makes this a commercial edition feature.

**Comment by knutwannheden** [comment]
> That sounds cool. I assume you would use `Meta::diff` to implement this, which then makes this a commercial edition feature.

Hmm, no, on second thought I don't see how you can use `Meta::diff` for this, unless the `Meta` snapshot from before applying the migration is still available. But after using `Meta::apply` one could of course use `Meta::diff` to "compute" the corresponding undo-migration.

**Comment by lukaseder** [comment]
> Hmm, no, on second thought I don't see how you can use Meta::diff for this, unless the Meta snapshot from before applying the migration is still available. But after using Meta::apply one could of course use Meta::diff to "compute" the corresponding undo-migration.

`Meta::diff` sounds like a quick win at first, but it will suffer from the same limitations in the unapply case as it does in the apply case.

No, I thought of interpreting individual statements, but applying the inverse semantics instead. Or, in other words, translating each statement to its inverse, prior to sending it to the interpreter. We can probably reach 90% of what's possible with "forward" interpretation without too much effort.

**Comment by lukaseder** [comment]
There are a few challenges I've discovered so far. I'll list them here and edit the list later:

**Irreversible statements**

Not only DML, but also some DDL statements are irreversible, because they change a value without the need of specifying the previous value. Some examples

- `COMMENT` statements
- `ALTER TABLE .. ALTER .. SET DATA TYPE` statement
- `ALTER TABLE .. ALTER .. SET DEFAULT` or `DROP DEFAULT` statement
- `ALTER SEQUENCE` statement, modifying sequence flags
- `DROP` statements can only be reversed if the statement references the complete meta data, not e.g. plain SQL meta data, which is not very likely in most scenarios.

Some altering statements are reversible, e.g.:

- Renaming usually involves both names, before and after
- `DROP NOT NULL` and `SET NOT NULL` can be assumed to be applied only to invert the nullability (i.e. `SET NOT NULL` will not be called on an already `NOT NULL` column. Some dialects don't even support this)

I will give this another thought, but perhaps this `unapply()` method is not a very good idea on this low level (on a per statement basis), unless we actually kept track of the entire `Meta` history as you suggested here: https://github.com/jOOQ/jOOQ/issues/9488#issuecomment-548742144
